Deprecated r_frame_rate, dropped in libavformat-56
Deprecated avcodec_encode_video, dropped in libavformat-56

--- modules/highgui/src/cap_ffmpeg_impl.hpp~	2013-12-27 18:09:18.000000000 +0100
+++ modules/highgui/src/cap_ffmpeg_impl.hpp	2014-11-06 13:59:49.977830576 +0100
@@ -769,7 +769,9 @@
     case CV_FFMPEG_CAP_PROP_FRAME_HEIGHT:
         return (double)frame.height;
     case CV_FFMPEG_CAP_PROP_FPS:
-#if LIBAVCODEC_BUILD > 4753
+#if LIBAVFORMAT_VERSION_MAJOR > 55
+        return av_q2d(video_st->avg_frame_rate);
+#elif LIBAVCODEC_BUILD > 4753
         return av_q2d(video_st->r_frame_rate);
 #else
         return (double)video_st->codec.frame_rate
@@ -817,7 +819,11 @@
 
 double CvCapture_FFMPEG::get_fps()
 {
+#if LIBAVFORMAT_VERSION_MAJOR > 55
+    double fps = r2d(ic->streams[video_stream]->avg_frame_rate);
+#else
     double fps = r2d(ic->streams[video_stream]->r_frame_rate);
+#endif
 
 #if LIBAVFORMAT_BUILD >= CALC_FFMPEG_VERSION(52, 111, 0)
     if (fps < eps_zero)
@@ -1203,6 +1209,25 @@
     return st;
 }
 
+#if LIBAVFORMAT_VERSION_MAJOR > 55
+/* replacement for deprecated avcodec_encode_video */
+static int encode_frame(AVCodecContext *c, AVFrame *frame)
+{
+    AVPacket pkt = { 0 };
+    int ret, got_output;
+
+    av_init_packet(&pkt);
+    av_init_packet(&pkt);
+    ret = avcodec_encode_video2(c, &pkt, frame, &got_output);
+    if (ret < 0)
+        return ret;
+
+    ret = pkt.size;
+    av_free_packet(&pkt);
+    return ret;
+}
+#endif
+
 static const int OPENCV_NO_FRAMES_WRITTEN_CODE = 1000;
 
 static int icv_av_write_frame_FFMPEG( AVFormatContext * oc, AVStream * video_st, uint8_t * outbuf, uint32_t outbuf_size, AVFrame * picture )
@@ -1233,7 +1258,11 @@
         ret = av_write_frame(oc, &pkt);
     } else {
         /* encode the image */
+#if LIBAVFORMAT_VERSION_MAJOR > 55
+        out_size = encode_frame(c, picture);
+#else
         out_size = avcodec_encode_video(c, outbuf, outbuf_size, picture);
+#endif
         /* if zero size, it means the image was buffered */
         if (out_size > 0) {
             AVPacket pkt;
